OSDOCS-19890 created zero-trust assembly/modules for multi-cluster (#112274)
openshift/openshift-docs: modules/zero-trust-manager-configure-spire-mesh-multi-cluster.adoc

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-configure-spire-mesh-multi-cluster_{context}"]
= Configuring {SMProductName} for multi-cluster {spire-full} integration

[role="_abstract"]
Configure {SMProductName} on each cluster with federation settings, East-West Gateways, and Remote Secrets to enable cross-cluster service communication by using SPIRE-issued certificates.

.Prerequisites

> * You have deployed {spire-full} with federation for multi-cluster integration. For more information, see "Deploying {spire-full} with federation for multi-cluster integration".

Reviewer: ```suggestion
* You deployed {spire-full} with federation for multi-cluster integration. For more information, see "Deploying {spire-full} with federation for multi-cluster integration".
```
past perfect tense is not needed here; remove for minimalism (not a merge blocker)

.Procedure

. Verify that the federation route is created on Cluster A by running the following command:
+
[source,terminal]
----
$ oc get route -n ${ZTWIM_NS} --kubeconfig="${CLUSTER_A_KUBECONFIG}" | grep federation
----

. Verify that the federation route is created on Cluster B by running the following command:
+
[source,terminal]
----
$ oc get route -n ${ZTWIM_NS} --kubeconfig="${CLUSTER_B_KUBECONFIG}" | grep federation
----

> . Create the `ClusterFederatedTrustDomain` CRs:

Reviewer: ```suggestion
```
don't need this IMHO

> .. On Cluster A, create a `ClusterFederatedTrustDomain` pointing to Cluster B by running the following command:

Reviewer: ```suggestion
. On Cluster A, create a `ClusterFederatedTrustDomain` object pointing to Cluster B by running the following command:
```

+
[source,terminal]
----

> $ oc apply --kubeconfig="${CLUSTER_A_KUBECONFIG}" -f - <<EOF

Reviewer: commands should still be separated from the YAML

apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterFederatedTrustDomain
metadata:
  name: federation-to-cluster-b
spec:
  trustDomain: ${CLUSTER_B_TRUST_DOMAIN}
  bundleEndpointURL: ${FEDERATION_ENDPOINT_B}
  bundleEndpointProfile:
    type: https_spiffe
    endpointSPIFFEID: spiffe://${CLUSTER_B_TRUST_DOMAIN}/spire/server
EOF
----

> .. On Cluster B, create a `ClusterFederatedTrustDomain` pointing to Cluster A by running the following command:

Reviewer: ```suggestion
. On Cluster B, create a `ClusterFederatedTrustDomain` pointing to Cluster A by running the following command:
```

  spire-server bundle list -socketPath /tmp/spire-server/private/api.sock -format spiffe 2>&1 | head -5
----
+
The output must show `${CLUSTER_B_TRUST_DOMAIN}` with its public keys.
+
.Example output
+
[source,text]
----
{
  "trust_domains": {
    "${CLUSTER_B_TRUST_DOMAIN}": {
      "keys": [
----

. Verify that the SPIRE Server on Cluster B has Cluster A's trust bundle by running the following command:
+
[source,terminal,subs="+quotes,verbatim"]
----
$ oc exec --kubeconfig="${CLUSTER_B_KUBECONFIG}" -n ${ZTWIM_NS} spire-server-0 -c spire-server -- \
  spire-server bundle list -socketPath /tmp/spire-server/private/api.sock -format spiffe 2>&1 | head -5
----
+
The output must show `${CLUSTER_A_TRUST_DOMAIN}` with its public keys.
+
.Example output
+
[source,text]
----
{
  "trust_domains": {
    "${CLUSTER_A_TRUST_DOMAIN}": {
      "keys": [
----

. Verify that the federation endpoint on Cluster A is reachable by running the following command:
+
[source,terminal]
----
$ curl -sk "${FEDERATION_ENDPOINT_A}" | python3 -c "import sys,json; print(f'Keys: {len(json.load(sys.stdin).get(\"keys\",[]))}')"
----
+

> The output must show at least two keys, one x509-svid and one jwt-svid.

Reviewer: ```suggestion
The output must show at least one `x509-svid` key and one `jwt-svid` key.
```

. Verify that the federation endpoint on Cluster B is reachable by running the following command:
+
[source,terminal]
----
$ curl -sk "${FEDERATION_ENDPOINT_B}" | python3 -c "import sys,json; print(f'Keys: {len(json.load(sys.stdin).get(\"keys\",[]))}')"
----
+

> The output must show at least two keys, one x509-svid and one jwt-svid.

Reviewer: ```suggestion
The output must show at least one `x509-svid` key and one `jwt-svid` key.
```

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-configure-spire-mesh-multi-cluster_{context}"]
= Configuring {SMProductName} for multi-cluster {spire-full} integration

[role="_abstract"]
Configure {SMProductName} on each cluster with federation settings, East-West Gateways, and Remote Secrets to enable cross-cluster service communication by using SPIRE-issued certificates.

.Prerequisites

* You have completed the "Deploying {spire-full} with federation for multi-cluster integration" steps.

.Procedure

. Enable federation on `SpireServer`:

.. Patch the `SpireServer` CR on both clusters to enable the federation endpoint:
+
[source,terminal]
----
//Cluster A

> $ oc patch spireserver cluster --kubeconfig="${CLUSTER_A_KUBECONFIG}" --type=merge -p '{

Reviewer 2: **CRITICAL: Federation schema is wrong — `enabled`, `address`, `port` do NOT exist on FederationConfig**

The federation patch uses non-existent fields:

```yaml
"federation": {
    "enabled": true,           # DOES NOT EXIST
    "bundleEndpoint": {
        "address": "0.0.0.0",  # DOES NOT EXIST
        "port": 8443           # DOES NOT EXIST
    }
}
```

The actual `FederationConfig` type ([source](https://github.com/openshift/zero-trust-workload-identity-manager/blob/main/api/v1alpha1/spire_server_config_types.go#L119)):

```go
type FederationConfig struct {
    BundleEndpoint BundleEndpointConfig `json:"bundleEndpoint"`
    FederatesWith  []FederatesWithConfig `json:"federatesWith,omitempty"`
    ManagedRoute   string `json:"managedRoute,omitempty"` // default: "true"
}
```

And `BundleEndpointConfig` ([source](https://github.com/openshift/zero-trust-workload-identity-manager/blob/main/api/v1alpha1/spire_server_config_types.go#L140)):

```go
// The federation endpoint is exposed on 0.0.0.0:8443  <-- hardcoded, not configurable
type BundleEndpointConfig struct {
    Profile     BundleEndpointProfile `json:"profile"`       // default: https_spiffe
    RefreshHint int32                 `json:"refreshHint,omitempty"` // default: 300
    HttpsWeb    *HttpsWebConfig       `json:"httpsWeb,omitempty"`    // only for https_web
}
```

Key points:
- **No `enabled` field** — federation is enabled by the presence of the `federation` object itself
- **No `address` or `port`** — the endpoint is hardcoded to `0.0.0.0:8443` per the type comment
- **`profile` defaults to `https_spiffe`**

**Correct federation config:**
```yaml
federation:
    bundleEndpoint:
        profile: https_spiffe
```

This same error appears in the inline SpireServer CR in `deploy-spire-multi-cluster.adoc` as well.

Author: Done.

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-configure-spire-mesh-multi-cluster_{context}"]
= Configuring {SMProductName} for multi-cluster {spire-full} integration

[role="_abstract"]

> Configure {SMProductName} on each cluster with federation settings, East-West Gateways, and Remote Secrets to enable cross-cluster service communication by using SPIRE-issued certificates.

Reviewer 3: ```suggestion
Configure {SMProductName} on each cluster with federation settings, east-west gateways, and remote secrets to enable cross-cluster service communication by using SPIRE-issued certificates.
```
Unless those terms are capitalized for some reason in this context, I don't think they need to be capitalized.

----

. Create the `ClusterFederatedTrustDomain` CRs:

.. On Cluster A, create a `ClusterFederatedTrustDomain` pointing to Cluster B by running the following command:
+
[source,terminal]
----
$ oc apply --kubeconfig="${CLUSTER_A_KUBECONFIG}" -f - <<EOF
apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterFederatedTrustDomain
metadata:
  name: federation-to-cluster-b
spec:
  trustDomain: ${CLUSTER_B_TRUST_DOMAIN}
  bundleEndpointURL: ${FEDERATION_ENDPOINT_B}
  bundleEndpointProfile:
    type: https_spiffe
    endpointSPIFFEID: spiffe://${CLUSTER_B_TRUST_DOMAIN}/spire/server
EOF
----

.. On Cluster B, create a `ClusterFederatedTrustDomain` pointing to Cluster A by running the following command:
+
[source,terminal]
----
$ oc apply --kubeconfig="${CLUSTER_B_KUBECONFIG}" -f - <<EOF
apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterFederatedTrustDomain
metadata:
  name: federation-to-cluster-a
spec:
  trustDomain: ${CLUSTER_A_TRUST_DOMAIN}
  bundleEndpointURL: ${FEDERATION_ENDPOINT_A}
  bundleEndpointProfile:
    type: https_spiffe
    endpointSPIFFEID: spiffe://${CLUSTER_A_TRUST_DOMAIN}/spire/server
EOF
----

.Verification

> . Verify that the SPIRE Server on Cluster A has Cluster B's trust bundle by running the following command:

Reviewer 3: ```suggestion
. Verify that the SPIRE Server on Cluster A has the trust bundle from Cluster B by running the following command:
```
Avoid contractions.

+
[source,terminal,subs="+quotes,verbatim"]
----
$ oc exec --kubeconfig="${CLUSTER_A_KUBECONFIG}" -n ${ZTWIM_NS} spire-server-0 -c spire-server -- \
  spire-server bundle list -socketPath /tmp/spire-server/private/api.sock -format spiffe 2>&1 | head -5
----
+
The output must show `${CLUSTER_B_TRUST_DOMAIN}` with its public keys.
+
.Example output
+
[source,text]
----
{
  "trust_domains": {
    "${CLUSTER_B_TRUST_DOMAIN}": {
      "keys": [
----

> . Verify that the SPIRE Server on Cluster B has Cluster A's trust bundle by running the following command:

Reviewer 3: ```suggestion
. Verify that the SPIRE Server on Cluster B has the trust bundle from Cluster A by running the following command:
```

//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-configure-spire-mesh-multi-cluster_{context}"]
= Configuring {SMProductName} for multi-cluster {spire-full} integration

[role="_abstract"]
Configure {SMProductName} on each cluster with federation settings, east-west gateways, and Remote Secrets to enable cross-cluster service communication by using SPIRE-issued certificates.

.Prerequisites

* You have deployed SPIRE with federation for multi-cluster integration. For more information, see "Deploying {spire-full} with federation on both clusters".

.Procedure

. Verify that the federation routes are created on Cluster A by running the following command:
+
[source,terminal]
----
$ oc get route -n ${ZTWIM_NS} --kubeconfig="${CLUSTER_A_KUBECONFIG}" | grep federation
----

. Verify that the federation routes are created on Cluster B by running the following command:
+
[source,terminal]
----
$ oc get route -n ${ZTWIM_NS} --kubeconfig="${CLUSTER_B_KUBECONFIG}" | grep federation
----

. Create the `ClusterFederatedTrustDomain` CRs:

> .. On Cluster A, create a `ClusterFederatedTrustDomain` pointing to Cluster B:

Reviewer 4: ```suggestion
.. On Cluster A, create a `ClusterFederatedTrustDomain` CR pointing to Cluster B:
```

----

.. On Cluster B, create a `ClusterFederatedTrustDomain` pointing to Cluster A:

... Create a YAML file that defines the `ClusterFederatedTrustDomain` CR on Cluster B:
+
[source,yaml]
----
apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterFederatedTrustDomain
metadata:
  name: federation-to-cluster-a
spec:
  trustDomain: ${CLUSTER_A_TRUST_DOMAIN}
  bundleEndpointURL: ${FEDERATION_ENDPOINT_A}
  bundleEndpointProfile:
    type: https_spiffe
    endpointSPIFFEID: spiffe://${CLUSTER_A_TRUST_DOMAIN}/spire/server
----

... Apply the YAML file on Cluster B by running the following command:
+
[source,terminal]
----
$ oc apply --kubeconfig="${CLUSTER_B_KUBECONFIG}" -f <filename>
----

.Verification

. Verify that the SPIRE Server on Cluster A has the trust bundle from Cluster B by running the following command:
+
[source,terminal,subs="+quotes,verbatim"]
----
$ oc exec --kubeconfig="${CLUSTER_A_KUBECONFIG}" -n ${ZTWIM_NS} spire-server-0 -c spire-server -- \
  spire-server bundle list -socketPath /tmp/spire-server/private/api.sock -format spiffe 2>&1 | head -5
----
+
The output must show public keys for `${CLUSTER_B_TRUST_DOMAIN}`.
+

> .Example output

Reviewer 4: @wgabor0427 I think you need to remove the `+` here to clear this error.

+
[source,text]
----
{
  "trust_domains": {
    "${CLUSTER_B_TRUST_DOMAIN}": {
      "keys": [
----

. Verify that the SPIRE Server on Cluster B has the trust bundle from Cluster A by running the following command:
+
[source,terminal,subs="+quotes,verbatim"]
----
$ oc exec --kubeconfig="${CLUSTER_B_KUBECONFIG}" -n ${ZTWIM_NS} spire-server-0 -c spire-server -- \
  spire-server bundle list -socketPath /tmp/spire-server/private/api.sock -format spiffe 2>&1 | head -5
----
+
The output must show public keys for `${CLUSTER_A_TRUST_DOMAIN}`.
+

> .Example output

Reviewer 4: @wgabor0427 I think you need to remove the + here to clear this error.

+
[source,text]
----
{
  "trust_domains": {
    "${CLUSTER_A_TRUST_DOMAIN}": {
      "keys": [
----

. Verify that the federation endpoint on Cluster A is reachable by running the following command:
+
[source,terminal]
----
$ curl -sk "${FEDERATION_ENDPOINT_A}" | python3 -c "import sys,json; print(f'Keys: {len(json.load(sys.stdin).get(\"keys\",[]))}')"
----
+
The output must show at least two keys, one x509-svid and one jwt-svid.
+

> .Example output

Reviewer 4: @wgabor0427 I think you need to remove the + here to clear this error.

+
[source,text]
----
Keys: 2
----

. Verify that the federation endpoint on Cluster B is reachable by running the following command:
+
[source,terminal]
----
$ curl -sk "${FEDERATION_ENDPOINT_B}" | python3 -c "import sys,json; print(f'Keys: {len(json.load(sys.stdin).get(\"keys\",[]))}')"
----
+
The output must show at least two keys, one x509-svid and one jwt-svid.
+

> .Example output

Reviewer 4: @wgabor0427 I think you need to remove the + here to clear this error.

// Module included in the following assemblies:
//
// * security/zero_trust_workload_identity_manager/zero-trust-manager-mesh-integration-multi-cluster.adoc.

:_mod-docs-content-type: PROCEDURE
[id="zero-trust-manager-configure-spire-mesh-multi-cluster_{context}"]
= Configuring {SMProductName} for multi-cluster {spire-full} integration

[role="_abstract"]
Configure {SMProductName} on each cluster with federation settings, East-West Gateways, and Remote Secrets to enable cross-cluster service communication by using SPIRE-issued certificates.

.Prerequisites

> * You have completed the "Deploying {spire-full} with federation for multi-cluster integration" steps.

Reviewer 5: reword to something like. Feel free to reword it to what makes most sense.
```suggestion
* You have deployed {spire-full} with federation for multi-cluster integration. For more information, see "Deploying {spire-full} with federation for multi-cluster integration".
```

.Procedure

> . Verify that the federation routes were created by running the following command:

Reviewer 5: ```suggestion
. Verify that the federation routes are created by running the following command:
```

done
----

. Create the `ClusterFederatedTrustDomain` CRs:

.. On Cluster A, create a `ClusterFederatedTrustDomain` pointing to Cluster B by running the following command:
+
[source,terminal]
----
$ oc apply --kubeconfig="${CLUSTER_A_KUBECONFIG}" -f - <<EOF
apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterFederatedTrustDomain
metadata:
  name: federation-to-cluster-b
spec:
  trustDomain: ${CLUSTER_B_TRUST_DOMAIN}
  bundleEndpointURL: ${FEDERATION_ENDPOINT_B}
  bundleEndpointProfile:
    type: https_spiffe
    endpointSPIFFEID: spiffe://${CLUSTER_B_TRUST_DOMAIN}/spire/server
EOF
----

.. On Cluster B, create a `ClusterFederatedTrustDomain` pointing to Cluster A by running the following command:
+
[source,terminal]
----
$ oc apply --kubeconfig="${CLUSTER_B_KUBECONFIG}" -f - <<EOF
apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterFederatedTrustDomain
metadata:
  name: federation-to-cluster-a
spec:
  trustDomain: ${CLUSTER_A_TRUST_DOMAIN}
  bundleEndpointURL: ${FEDERATION_ENDPOINT_A}
  bundleEndpointProfile:
    type: https_spiffe
    endpointSPIFFEID: spiffe://${CLUSTER_A_TRUST_DOMAIN}/spire/server
EOF
----

.Verification

> . Check that each SPIRE Server has the remote cluster's trust bundle by running the following command:

Reviewer 5: ```suggestion
. Verify that each SPIRE Server has the remote cluster's trust bundle by running the following command:
```

+
[source,terminal,subs="+quotes,verbatim"]
----
$ oc exec --kubeconfig="${CLUSTER_A_KUBECONFIG}" -n ${ZTWIM_NS} spire-server-0 -c spire-server -- \
  spire-server bundle list -socketPath /tmp/spire-server/private/api.sock -format spiffe 2>&1 | head -5
----
+

> The output should show `apps.cluster-b.example.com` with its public keys.

Reviewer 5: ```suggestion
The output must show `apps.cluster-b.example.com` with its public keys.
```

Reviewer 5: Verification steps should have a sample output.

+
[source,terminal,subs="+quotes,verbatim"]
----
$ oc exec --kubeconfig="${CLUSTER_B_KUBECONFIG}" -n ${ZTWIM_NS} spire-server-0 -c spire-server -- \
  spire-server bundle list -socketPath /tmp/spire-server/private/api.sock -format spiffe 2>&1 | head -5
----
+

> The output should show `apps.cluster-a.example.com` with its public keys.

Reviewer 5: ```suggestion
The output must show `apps.cluster-a.example.com` with its public keys.
```

Reviewer 5: Verification steps should have a sample output.

> . Verify that the federation endpoints exist by running the following commands:

Reviewer 5: split this into 2 commands, for A and B

+
[source,terminal]
----
$ curl -sk "${FEDERATION_ENDPOINT_A}" | python3 -c "import sys,json; print(f'Keys: {len(json.load(sys.stdin).get(\"keys\",[]))}')"
----
+
[source,terminal]
----
$ curl -sk "${FEDERATION_ENDPOINT_B}" | python3 -c "import sys,json; print(f'Keys: {len(json.load(sys.stdin).get(\"keys\",[]))}')"
----
+

> Both should return at least two keys, one x509-svid and one jwt-svid.

Reviewer 5: sample output?
